core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java

```diff
@@ -109,18 +112,34 @@ protected abstract class BaseEqualityDeltaWriter implements Closeable {
     private final StructProjection structProjection;
     private RollingFileWriter dataWriter;
     private RollingEqDeleteWriter eqDeleteWriter;
-    private SortedPosDeleteWriter<T> posDeleteWriter;
+    private FileWriter<PositionDelete<T>, DeleteWriteResult> posDeleteWriter;
     private Map<StructLike, PathOffset> insertedRowMap;
 
     protected BaseEqualityDeltaWriter(StructLike partition, Schema schema, Schema deleteSchema) {
+      this(partition, schema, deleteSchema, DeleteGranularity.PARTITION);
+    }
+
+    protected BaseEqualityDeltaWriter(
+        StructLike partition,
+        Schema schema,
+        Schema deleteSchema,
+        DeleteGranularity deleteGranularity) {
       Preconditions.checkNotNull(schema, "Iceberg table schema cannot be null.");
       Preconditions.checkNotNull(deleteSchema, "Equality-delete schema cannot be null.");
       this.structProjection = StructProjection.create(schema, deleteSchema);
 
       this.dataWriter = new RollingFileWriter(partition);
       this.eqDeleteWriter = new RollingEqDeleteWriter(partition);
       this.posDeleteWriter =
-          new SortedPosDeleteWriter<>(appenderFactory, fileFactory, format, partition);
+          new SortingPositionOnlyDeleteWriter<>(
+              () ->
+                  appenderFactory.newPosDeleteWriter(
+                      partition == null
```

Comment:
If I am not mistaken, this line should check `spec.isUnpartitioned() || partition == null` to avoid weird issues we discovered in #7685. We also have to pass `spec` with `partition` below in `newOutputFile`.

I'd also consider moving this logic into a separate method for better readability. Up to you, though.

```
private EncryptedOutputFile newOutputFile(StructLike partition) {
  if (spec.isUnpartitioned() || partition == null) {
    return fileFactory.newOutputFile();
  } else {
    return fileFactory.newOutputFile(spec, partition);
  }
}
```

Reply:
Done.
Thanks for the catch!
I just refactored the previous, but your suggestion makes sense